pkg/cloud/aws/services/ec2/utils.go

```diff
@@ -0,0 +1,139 @@
+package ec2
+
+import (
+	"fmt"
+	"sort"
+
+	providerconfigv1 "sigs.k8s.io/cluster-api-provider-aws/pkg/apis/awsproviderconfig/v1alpha1"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/golang/glog"
+)
+
+type instanceList []*ec2.Instance
+
+func (il instanceList) Len() int {
+	return len(il)
+}
+
+func (il instanceList) Swap(i, j int) {
+	il[i], il[j] = il[j], il[i]
+}
+
+func (il instanceList) Less(i, j int) bool {
+	if il[i].LaunchTime == nil && il[j].LaunchTime == nil {
+		// No idea what to do here, should not be possible, so keep the order
+		return false
+	}
+	if il[i].LaunchTime != nil && il[j].LaunchTime == nil {
+		return false
+	}
+	if il[i].LaunchTime == nil && il[j].LaunchTime != nil {
+		return true
+	}
+	if (*il[i].LaunchTime).After(*il[j].LaunchTime) {
+		return true
+	}
+	return false
+}
+
+// SortInstances will sort a list of instance based on an instace launch time
+// from the newest to the latest.
+// This function should only be called with running instances, not those which are stopped or
+// terminated.
+func SortInstances(instances []*ec2.Instance) {
+	sort.Sort(instanceList(instances))
+}
+
+func buildEC2Filters(inputFilters []providerconfigv1.Filter) []*ec2.Filter {
+	filters := make([]*ec2.Filter, len(inputFilters))
+	for i, f := range inputFilters {
+		values := make([]*string, len(f.Values))
+		for j, v := range f.Values {
+			values[j] = aws.String(v)
+		}
+		filters[i] = &ec2.Filter{
+			Name:   aws.String(f.Name),
+			Values: values,
+		}
+	}
+	return filters
+}
+
+func (ir *instancesRuntime) getSecurityGroupsIDs(securityGroups []providerconfigv1.AWSResourceReference) ([]*string, error) {
```

Comment:
The separation seems a bit odd to me, shouldn't this be in the actuator package?. We are putting this on a separate and create a mirror struct `LaunchConfig` that is still coupled with providerconfigv1. As mentioned above If any I would create funcs for the logic for getting the final input (e.g getAMIID) based on the user input so that can be unit tested. Then if we think there's a benefit from having a `LaunchInstance` func it could just receive `RunInstancesInput` cc @frobware

Reply:
I don't want the actuator to be aware of what's inside the machine object unless it's necessary. The `services/ec2` (or we can rename it to something else) is the only package that is responsible for launching an aws instance. So only this package should know what fields of the machine provider config are needed to run an instance. So it only make sense to put all the logic you are describing in the comment here.